capi-configuration/0_capi-releases.md

```diff
@@ -0,0 +1,205 @@
+# Releases as gitops managed configmaps
+
+## Problem Statement
+
+We need to ...
+1. be able to define Giant Swarm releases as sets of configuration.
+2. be able to create Giant Swarm clusters in a repeatable and reliable way.
+3. be able to upgrade Giant Swarm clusters in a repeatable way with `cluster-api`.
+4. be able to default Giant Swarm clusters through all our user interfaces consistently.
+
+We want to ...
+1. decouple managed applications from strict release flows.
+
+## Git-Ops approach
+
+We have already taken steps to manage configuration for our components in management clusters through open source gitops tools.
+This has set a precedent for gitops management for configuration in other areas as well.
+
+Our customers generally desire to work with an `Infrastructure As Code` approach which is extended by currently available open source gitops tooling.
+
+**This RFC will exclusively deal with the impact of using a gitops driven approach in the following sections.**
+
+## Configuration structure
+
+There are many different possible ways to structure the configuration which is used to create the clusters.
+The goal here is to provide one way that we think _can_ work and then iterating on it - we do not want to box ourselves in.
+
+### Cluster-api fundamental structure
+
+There are some links between and some fields in `cluster-api` custom resources which are essential.
+These fields can not be inferred from a webhook without context knowledge.
+It is therefore necessary to supply these values _immediately_ when the custom resource is created (e.g. `kubectl-gs` or `happa` needs to fill them).
+We therefore define `fundamental templates` as minimal templates which have to be submitted to create a working cluster.
+
+We can look at a valid `MachinePool` custom resource to determine its fundamental structure:
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachinePool
+metadata:
+  name: capa-mp-0 # This is taken from upstream, Giant Swarm would use an ID like `eump1`.
```

Comment:
Just for the completeness, shall we add the namespace? I'm asking because of the organization handling.

Reply:
Makes sense to me - will add namespace in all examples. Do you think the namespace should be set by clients? I am not sure how we currently determine this with organisations and so on.

I would guess the client (e.g. happa) sets the namespace.

Reply:
From what I know in firecracker, we do set the namespace by default with release AWS 16.0.0 and also validate that all CR's matching with the Cluster CR namespace if that is not correct. But only with new cluster creation.